### DISCUSSION TITLE
Sending a SSE:Event with leading Whitespace

### DISCUSSION THREAD

**Discussion by michaelfeil** [tokio-rs/axum#2146] [General]
Using https://docs.rs/axum/latest/axum/response/sse/struct.Event.html, I am trying to send a SSE, where the `data:` field of SSe is a bit special, it starts with a single leading whitespace.

Sadly, I cannot modify this requirement, its in the openai streaming protocol https://github.com/openai/openai-python/blob/b82a3f7e4c462a8a10fa445193301a3cefef9a4a/openai/api_requestor.py#L102
Clients would expect as stop token for the SSE,  python 3.9 would check for `b"data: [DONE]"` -> aka payload of event is `b" [DONE]"` -> `[32, 91, 68, 79, 78, 69, 93]` in utf-8 bytes as e.g. stop sequence.

```rust
let stop_seq = vec![32, 91, 68, 79, 78, 69, 93];
let stop_seq = std::str::from_utf8(&stop_seq).unwrap();
yield Ok(Event::default().data(stop_seq));
```

When trying this code above, I however yield the following sequence:
`[100, 97, 116, 97, 58, 32, 32, 91, 68, 79, 78, 69, 93]` for `Event` `data`  `vec![32, 91, 68, 79, 78, 69, 93];`
and 
`[100, 97, 116, 97, 58, 91, 68, 79, 78, 69, 93]` for `Event` `data`  `vec![91, 68, 79, 78, 69, 93];`

How can I add a leading whitespace? 
s

Same would go on how to add 
```rust
stream_token: Box<dyn Serialize + Send> = ..
yield Ok(Event::default().json_data(stream_token).unwrap());
```

FYI:
- Context with some visualizations: https://github.com/Preemo-Inc/text-generation-inference/pull/2
- Mozilla SSE docu contains leading whitespace for SSE: https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#examples
- This started all here https://github.com/tokio-rs/axum/pull/600 and here: https://github.com/tokio-rs/axum/blob/432289dd042d4f91af54dda9d9b2abaaae7d36be/axum/src/response/sse.rs#L540C1-L544C70 
   0 whitespace -> 0 whitespace
   1 whitespace -> 2 whitespaces for data:

**Comment by davidpdrsn** [comment]
I was gonna suggest that you include an additional space yourself with `.data(" foo")` but it seems we have a bug where that causes an additional space to be added here https://github.com/tokio-rs/axum/blob/main/axum/src/response/sse.rs#L362-L364

I can fix that tomorrow. Would that work for you?

**Reply by michaelfeil** [reply]
Would work for me! Thanks a lot for checking this out!

I think the PR was a mistake, the implementation is not conform to the standard, a leading space after data: should be ignored, most other libraries in Java, Python implement data field with white-space included as `b"data: "` to avoid this confusion. 
https://github.com/sysid/sse-starlette/blob/03e4fa1dc6dfbe63febb5fc8a1a48234e8a8f707/tests/test_sse.py#L14

https://html.spec.whatwg.org/multipage/server-sent-events.html#processField
 
_The following stream fires two identical events:_
```text
data:test

data: test
```
_This is because the space after the colon is ignored if present._

**Reply by davidpdrsn** [reply]
Yeah I see it in the spec as well https://html.spec.whatwg.org/multipage/server-sent-events.html#event-stream-interpretation

We also have this test specifically for the double space https://github.com/tokio-rs/axum/blob/main/axum/src/response/sse.rs#L543-L544

I suppose that means everything is working fine?

**Reply by michaelfeil** [reply]
No, just added some context. As you said earlier, it does not work, sending  `b"data: [DONE]"` is impossible with the current implementation.

**Reply by davidpdrsn** [reply]
Shouldn't that be fixed in the client? Then all severs benefit.

**Reply by michaelfeil** [reply]
I am already doing so. The protocol is many ways reverse engineered, e.g. https://github.com/openai/openai-python/pull/559
